haystack/core/pipeline/base.py

```diff
@@ -816,6 +816,107 @@ def _init_graph(self):
+    ):
+        """
+        Distributes the output of a Component to the next Components that need it.
+
+        This also updates the queues that keep track of which Components are ready to run and which are waiting for input.
+
+        :param component_name: Name of the Component that created the output
+        :param component_result: The output of the Component
+        :paramt inputs_by_component: The current state of the inputs divided by Component name
+        :param to_run: Queue of Components to run
+        :param waiting_for_input: Queue of Components waiting for input
+
+        :return: The updated output of the Component without the keys that were distributed to other Components
+        """
+        # We keep track of which keys to remove from res at the end of the loop.
+        # This is done after the output has been distributed to the next components, so that
+        # we're sure all components that need this output have received it.
+        to_remove_from_res = set()
+
+        for _, receiver_name, connection in self.graph.edges(nbunch=component_name, data=True):
+            sender_socket: OutputSocket = connection["from_socket"]
+            receiver_socket: InputSocket = connection["to_socket"]
+
+            if sender_socket.name not in component_result:
+                # This output wasn't created by the sender, nothing we can do
+                continue
+
+            if receiver_name not in inputs_by_component:
+                inputs_by_component[receiver_name] = {}
+
+            # We'll remove this key from the output at the end of the loop
+            to_remove_from_res.add(sender_socket.name)
+
+            value = component_result[sender_socket.name]
+
+            if receiver_socket.is_variadic:
+                # Variadic inputs are always lists
+                if receiver_socket.name not in inputs_by_component[receiver_name]:
+                    # Create the list if it doesn't exist
+                    inputs_by_component[receiver_name][receiver_socket.name] = []
+                inputs_by_component[receiver_name][receiver_socket.name].append(value)
+            else:
+                inputs_by_component[receiver_name][receiver_socket.name] = value
+
+            receiver = self.graph.nodes[receiver_name]["instance"]
+            pair = (receiver_name, receiver)
+
+            is_greedy = getattr(receiver, "__haystack_is_greedy__", False)
+            if receiver_socket.is_variadic and is_greedy:
+                # If the receiver is greedy, we can run it right away.
+                # First we remove it from the status lists it's in if it's there or we risk running it multiple times.
+                if pair in to_run:
+                    to_run.remove(pair)
+                if pair in waiting_for_input:
+                    waiting_for_input.remove(pair)
+                to_run.append(pair)
+            elif pair not in waiting_for_input and pair not in to_run:
```

Comment:
I don't follow how this condition is mutually exclusive of the previous one? Why aren't we performing the same checks above as well?

Reply:
If it goes in the first `if` branch it won't go in the second and viceversa. I decided to use an `elif` to avoid checking a condition that I know can't be `True` if the first is.

I can make `elif` an `if `and functionality won't change anyway.

Reply:
Let's change that - it will then telegraph that the first `if` clause is performing special-case behaviour. Might not change the functionality but will improve readability and communicate the intent better.